spring-cloud-gateway-server-mvc/src/main/java/org/springframework/cloud/gateway/server/mvc/handler/ProxyExchangeHandlerFunction.java

```diff
@@ -84,14 +84,12 @@ private void init() {
 	@Override
 	public ServerResponse handle(ServerRequest serverRequest) {
 		URI uri = uriResolver.apply(serverRequest);
-		boolean encoded = containsEncodedQuery(serverRequest.uri(), serverRequest.params());
```

Comment:
Why is this check being removed?  Can we assume URI is always encoded?

Reply:
I am not sure why this check was added in the first place. In my opinion, spring-cloud-gateway should not encode or decode parameters or URI parts, but leave them untouched. A similar change is also proposed in https://github.com/spring-cloud/spring-cloud-gateway/pull/3437/files

Reply:
I think the intent may be if the incoming URI is already encoded, keep the encoding...otherwise don't.  Not working that way though, I'm seeing the request is decoded by the time it reaches the backend.

The key is handling replaceQueryParms piece below, may be we need to encode the request parms if the request is already encoded ?

something like

boolean encoded = containsEncodedQuery(serverRequest.uri(), serverRequest.params());

var parms = serverRequest.params();;

if(encoded) {
    //pseudo code
    parms = encode(serverRequest.params());
}

	URI url = UriComponentsBuilder.fromUri(serverRequest.uri())
				.scheme(uri.getScheme())
				.host(uri.getHost())
				.port(uri.getPort())
				.replaceQueryParams(parms))
				.build(encoded)
				.toUri();